Authenticate blinded payment paths

jkczyz opened the pull request:
When receiving a `PaymentContext` from a blinded payment, the context must be authenticated. Otherwise, the context can be forged and would appear within a `PaymentPurpose`. In order to authenticate a `PaymentContext`, an `Hmac` and a `Nonce` must be included along with it in `payment::ReceiveTlvs`. Upon receiving an HTLC over a blinded path, `ChannelManager` can then authenticate it.

This is also useful for preventing de-anonymization attacks where `payment::ReceiveTlvs` is forged to use a `PaymentSecret` from a BOLT11 invoice.

Fixes #3427

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -6205,6 +6207,16 @@ where
+								if let Some(payment_context) = payment_context.as_ref() {
+									if let Some((hmac, nonce)) = authentication {
+										if payment_context.verify_for_offer_payment(hmac, nonce, &self.inbound_payment_key).is_err() {
+											fail_htlc!(claimable_htlc, payment_hash);
```
Hmmm... we could, but it would require having access to the `ExpandedKey`. And that would leak into the public `onion_payment::peel_payment_onion` function's interface, too, IIUC. Maybe that is ok?

jkczyz replied:
Yeah, I'm inclined to update `NodeSigner` as you mentioned. Also, instead of touching `peel_payment_onion`, we could do authenticate a little earlier when parsing the `InboundOnionPayload`. I guess this is more what you meant? It also has a `NodeSigner` and the entire `ReceiveTlvs`.

jkczyz replied:
Updated accordingly.

jkczyz commented on lightning/src/blinded_path/payment.rs:
```diff
@@ -260,6 +264,8 @@ pub struct ReceiveTlvs {
 	/// Context for the receiver of this payment.
 	pub payment_context: PaymentContext,
+	/// An HMAC of `payment_context` along with a nonce used to construct it.
+	pub authentication: (Hmac<Sha256>, Nonce),
```
So still write as we are for `BlindedPaymentTlvs` but read it as a separate component of `BlindedPaymentTlvs::Receive`? We'd need to pass it through to `Router::create_blinded_payment_paths`, `BlindedPaymentPath::new`, etc. Though, I think that would mean we wouldn't need to modify `Bolt12RefundContext` since the secret would be included.

jkczyz replied:
Gotcha, yeah that would reduce the amount of code we'd need to touch.

jkczyz replied:
Updated to use an `Option`, which is `take`n at verification time.

jkczyz commented on lightning/src/offers/signer.rs:
```diff
@@ -46,6 +47,9 @@ const ASYNC_PAYMENT_ID_HMAC_INPUT: &[u8; 16] = &[6; 16];
 const PAYMENT_HASH_HMAC_INPUT: &[u8; 16] = &[7; 16];
 
+// HMAC input for `ReceiveTlvs`. The HMAC is used in `blinded_path::payment::PaymentContext`.
+const PAYMENT_TLVS_HMAC_INPUT: &[u8; 16] = &[8; 16];
```
Note that this needs to be updated as it conflicts with #3408.

jkczyz commented on lightning/src/sign/mod.rs:
```diff
@@ -832,7 +833,7 @@ pub trait NodeSigner {
 	/// This method must return the same value each time it is called.
 	///
 	/// [phantom node payments]: PhantomKeysManager
-	fn get_inbound_payment_key_material(&self) -> KeyMaterial;
```
Yeah, will add a separate commit given it touches a lot of tests.

jkczyz replied:
Removed now.

jkczyz commented on lightning/src/ln/msgs.rs:
```diff
@@ -2911,9 +2908,19 @@ impl<NS: Deref> ReadableArgs<(Option<PublicKey>, NS)> for InboundOnionPayload wh
+							return Err(DecodeError::InvalidValue);
+						}
+					} else {
+						return Err(DecodeError::InvalidValue);
```
Done.

valentinewallace replied:
I think we'll also fail backwards without blinding the error. That might give a hint that we're the recipient if combined with other information, not sure if also worth noting. I guess this would also mean that someone can create a blinded path terminating at us and force us to reveal that we are the last node in their inauthentic path 🤔 weird case though

jkczyz replied:
> I think we'll also fail backwards without blinding the error. That might give a hint that we're the recipient if combined with other information, not sure if also worth noting.

Would we need to use the original approach to prevent this?

> I guess this would also mean that someone can create a blinded path terminating at us and force us to reveal that we are the last node in their inauthentic path 🤔 weird case though

Not sure I understand this scenario. Don't they know we are the recipient? Or is it being revealed to someone else along the path?

valentinewallace replied:
> Would we need to use the original approach to prevent this?

The original approach should prevent this, or alternately add a new `DecodeError` or somehow bubble up from the `Read` method that the error needs to be blinded. 

> Not sure I understand this scenario. Don't they know we are the recipient? Or is it being revealed to someone else along the path?

Ah, it would be potentially revealed to the hop preceding us. I'm not sure this is worth addressing since it's a fairly narrow pre-0.1 case, but yeah just wanted to point it out.

jkczyz commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -10458,20 +10471,28 @@ where
 			payment_context,
+			authentication: None,
 		};
+		let nonce = Nonce::from_entropy_source(entropy);
+		let hmac = payee_tlvs.hmac_for_offer_payment(nonce, expanded_key);
+		payee_tlvs.authentication = Some((hmac, nonce));
```
I'll need to document how this should be set. But I'm wondering if we should pass the `NodeSigner` to `BlindedPaymentPath::new`, which already has an `EntropySource`. Then anyone creating a `BlindedPaymentPath` wouldn't need to do this explicitly. Drawback is we'd either need to pass both of those to `Router::create_blinded_payment_paths` or initialize `DefaultRouter` with a `NodeSigner` -- it already has an `EntropySource`. 

@TheBlueMatt Not sure if that is really preferable. Thoughts?

shaavan replied:
I was wondering if we could consider passing `expanded_key` as an input parameter to `Router::create_blinded_payment_paths`. This could potentially enable `DefaultRouter` to handle both `HMAC` and `nonce` creation directly within `create_blinded_payment_paths`, setting them appropriately in the process.

However, I noticed that the codebase doesn't seem to have any `pub` functions that take `expanded_key` as an input parameter, so this approach might not be feasible.

Curious to hear your thoughts on this!

jkczyz replied:
Ok, I took some inspiration from this and a discussion based on @shaavan's [comment](https://github.com/lightningdevkit/rust-lightning/pull/3435#pullrequestreview-2488795505) to make an `UnauthenticatedReceiveTlvs` with a method to convert to `ReceiveTlvs`. This way we can never have `ReceiveTlvs` in an invalid state since the `Option` is no longer used.

jkczyz commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -1416,6 +1435,60 @@ fn custom_tlvs_to_blinded_path() {
 }
 
+#[test]
+fn fails_receive_tlvs_authentication() {
```
@valentinewallace I cribbed this off some of the other tests here. Let me know if something looks off or if it could be done in a simpler manner.

vincenzopalazzo commented on lightning/src/blinded_path/payment.rs:
```diff
@@ -437,7 +424,7 @@ impl Readable for BlindedPaymentTlvs {
 			(14, features, (option, encoding: (BlindedHopFeatures, WithoutLength))),
 			(65536, payment_secret, option),
-			(65537, payment_context, (default_value, PaymentContext::unknown())),
+			(65537, payment_context, option),
```
Probably a silly question about the code here, but what happens if the `payment_context` is already set to `UnknownPaymentContext` in a running LDK node? 

Is it possible for this event to occur?

shaavan replied:
Let me give this a shot!

Since we are removing the variant in this PR, an upgraded LDK node won't "understand" the `payment_context` it encounters during a read. As a result, parsing the `BlindedPaymentTlvs` and, by extension, that `Blinded Payment Path` will fail. However, this should be fine because the introduction of `authentication` in this PR inherently breaks legacy compatibility, meaning the path wouldn't be usable regardless.

@jkczyz, hope I got this right! :)

jkczyz replied:
Yeah, when receiving a payment we'd fail to parse any TLVs using `PaymentContext::unknown()`. I believe this should be rare as it would only occur for `BlindedPaymentPath`s that preexisted `PaymentContext`. Additionally, the `PaymentContext::Unknown` variant cannot be created outside of the crate, so there's no way for a user to construct one. Any `BlindedPaymentPath` containing one could be parsed, however, since that data is encrypted in the path. But we never created those. It's only used when parsing the TLVs from incoming payments.

vincenzopalazzo reviewed (commented):
Concept ACK, I left a question for learning purposes :)

shaavan reviewed (commented):
Looks great on the first pass! 🚀  

**One minor suggestion:**

Since we shouldn't create or accept a `BlindedPaymentPath` without `ReceiveTlvs` authentication, how about adding a fail-safe in `Router::create_blinded_payment_path`? For example, a `debug_assert!()` or an `Err!()` that triggers if `receive_tlvs.authentication.is_none()`. This could help catch any incorrect usage early on.

jkczyz commented:
Squashed the fixups as per our offline conversation.

valentinewallace reviewed (commented):
Basically looks good on first pass, will take another look tomorrow!

valentinewallace commented:
Needs rebase

valentinewallace commented:
LGTM after rebase!

valentinewallace commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -1416,6 +1430,65 @@ fn custom_tlvs_to_blinded_path() {
+		amt_msat,
+	);
+
+	nodes[0].node.send_payment(payment_hash, RecipientOnionFields::spontaneous_empty(), PaymentId(payment_hash.0), route_params, Retry::Attempts(0)).unwrap();
+	check_added_monitors(&nodes[0], 1);
+
+	let mut events = nodes[0].node.get_and_clear_pending_msg_events();
+	assert_eq!(events.len(), 1);
+	let ev = remove_first_msg_event_to_node(&nodes[1].node.get_our_node_id(), &mut events);
+	let mut payment_event = SendEvent::from_event(ev);
+
+	nodes[1].node.handle_update_add_htlc(nodes[0].node.get_our_node_id(), &payment_event.msgs[0]);
+	check_added_monitors!(nodes[1], 0);
+	do_commitment_signed_dance(&nodes[1], &nodes[0], &payment_event.commitment_msg, true, true);
+	nodes[1].node.process_pending_htlc_forwards();
```
Oh, btw I think you might be able to replace this with something like:
```rust
	nodes[0].node.send_payment(payment_hash, RecipientOnionFields::spontaneous_empty(),
	PaymentId(payment_hash.0), route_params, Retry::Attempts(0)).unwrap();
	check_added_monitors(&nodes[0], 1);
	pass_along_route(&nodes[0], &[&[&nodes[1]]], amt_msat, payment_hash, payment_secret);
```
Taken from `do_one_hop_blinded_path`. But not a big deal.

jkczyz replied:
Hmmm... I see the following failure using that since the HTLC should fail, IIUC.

```
thread 'ln::blinded_payment_tests::fails_receive_tlvs_authentication' panicked at 'assertion failed: node_a.node.get_and_clear_pending_msg_events().is_empty()', lightning/src/ln/functional_test_utils.rs:2141:9
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:142:14
   2: core::panicking::panic
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/panicking.rs:48:5
   3: lightning::ln::functional_test_utils::do_commitment_signed_dance
             at ./src/ln/functional_test_utils.rs:2141:3
   4: lightning::ln::functional_test_utils::do_pass_along_path
             at ./src/ln/functional_test_utils.rs:2710:4
   5: lightning::ln::functional_test_utils::pass_along_path
             at ./src/ln/functional_test_utils.rs:2789:2
   6: lightning::ln::functional_test_utils::pass_along_route
             at ./src/ln/functional_test_utils.rs:2820:3
   7: lightning::ln::blinded_payment_tests::fails_receive_tlvs_authentication
             at ./src/ln/blinded_payment_tests.rs:1470:2
   8: lightning::ln::blinded_payment_tests::fails_receive_tlvs_authentication::{{closure}}
             at ./src/ln/blinded_payment_tests.rs:1434:1
   9: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
  10: core::ops::function::FnOnce::call_once
             at /rustc/4b91a6ea7258a947e59c6522cd5898e7c0a6a88f/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

valentinewallace replied:
Oh, right, the mentioned test method does include a failure but it actively fails the payment. Not gonna bother looking into it further lol.

jkczyz commented on lightning/src/ln/blinded_payment_tests.rs:
```diff
@@ -94,15 +94,6 @@ fn blinded_payment_path(
-	let expanded_key = keys_manager.get_inbound_payment_key();
-	let hmac = payee_tlvs.hmac_for_offer_payment(nonce, &expanded_key);
-	(hmac, nonce)
-}
```
Managed to move parts of that commit across the relevant commits. Required moving some commits around, though.

jkczyz commented:
Rebased

jkczyz commented:
Addressed the nits.

```diff
diff --git a/lightning/src/blinded_path/payment.rs b/lightning/src/blinded_path/payment.rs
index 3f3c3c768..e3a819271 100644
--- a/lightning/src/blinded_path/payment.rs
+++ b/lightning/src/blinded_path/payment.rs
@@ -624,6 +624,7 @@ impl Readable for PaymentConstraints {
 
 impl_writeable_tlv_based_enum_legacy!(PaymentContext,
        ;
+       // 0 for Unknown removed in version 0.1.
        (1, Bolt12Offer),
        (2, Bolt12Refund),
 );
diff --git a/lightning/src/ln/blinded_payment_tests.rs b/lightning/src/ln/blinded_payment_tests.rs
index 7d25c472b..c1fad65c1 100644
--- a/lightning/src/ln/blinded_payment_tests.rs
+++ b/lightning/src/ln/blinded_payment_tests.rs
@@ -1433,7 +1433,7 @@ fn fails_receive_tlvs_authentication() {
        let chan_upd = create_announced_chan_between_nodes_with_value(&nodes, 0, 1, 1_000_000, 0).0.contents;
 
        let amt_msat = 5000;
-       let (_, payment_hash, payment_secret) = get_payment_preimage_hash(&nodes[1], Some(amt_msat), None);
+       let (payment_preimage, payment_hash, payment_secret) = get_payment_preimage_hash(&nodes[1], Some(amt_msat), None);
        let payee_tlvs = UnauthenticatedReceiveTlvs {
                payment_secret,
                payment_constraints: PaymentConstraints {
@@ -1444,8 +1444,37 @@ fn fails_receive_tlvs_authentication() {
        };
        let nonce = Nonce([42u8; 16]);
        let expanded_key = chanmon_cfgs[1].keys_manager.get_inbound_payment_key();
-       let mut payee_tlvs = payee_tlvs.authenticate(nonce, &expanded_key);
+       let payee_tlvs = payee_tlvs.authenticate(nonce, &expanded_key);
+
+       let mut secp_ctx = Secp256k1::new();
+       let blinded_path = BlindedPaymentPath::new(
+               &[], nodes[1].node.get_our_node_id(), payee_tlvs, u64::MAX, TEST_FINAL_CLTV as u16,
+               &chanmon_cfgs[1].keys_manager, &secp_ctx
+       ).unwrap();
+
+       let route_params = RouteParameters::from_payment_params_and_value(
+               PaymentParameters::blinded(vec![blinded_path]),
+               amt_msat,
+       );
+
+       // Test authentication works normally.
+       nodes[0].node.send_payment(payment_hash, RecipientOnionFields::spontaneous_empty(), PaymentId(payment_hash.0), route_params, Retry::Attempts(0)).unwrap();
+       check_added_monitors(&nodes[0], 1);
+       pass_along_route(&nodes[0], &[&[&nodes[1]]], amt_msat, payment_hash, payment_secret);
+       claim_payment(&nodes[0], &[&nodes[1]], payment_preimage);
+
        // Swap in a different nonce to force authentication to fail.
+       let (_, payment_hash, payment_secret) = get_payment_preimage_hash(&nodes[1], Some(amt_msat), None);
+       let payee_tlvs = UnauthenticatedReceiveTlvs {
+               payment_secret,
+               payment_constraints: PaymentConstraints {
+                       max_cltv_expiry: u32::max_value(),
+                       htlc_minimum_msat: chan_upd.htlc_minimum_msat,
+               },
+               payment_context: PaymentContext::Bolt12Refund(Bolt12RefundContext {}),
+       };
+       let nonce = Nonce([43u8; 16]);
+       let mut payee_tlvs = payee_tlvs.authenticate(nonce, &expanded_key);
        payee_tlvs.authentication.1 = Nonce([0u8; 16]);
 
        let mut secp_ctx = Secp256k1::new();
```

jkczyz commented on lightning/src/ln/inbound_payment.rs:
```diff
@@ -129,7 +129,7 @@ fn min_final_cltv_expiry_delta_from_metadata(bytes: [u8; METADATA_LEN]) -> u16 {
 ///
-/// `keys` is generated by calling [`NodeSigner::get_inbound_payment_key_material`] and then
+/// `keys` is generated by calling [`NodeSigner::get_inbound_payment_key`] and then
 /// calling [`ExpandedKey::new`] with its result. It is recommended to cache this value and not
```
Sorry, that was an oversight on my part. Completely missed the comment somehow.

jkczyz commented:
Fixed docs as per missed comment.

```diff
diff --git a/lightning/src/ln/inbound_payment.rs b/lightning/src/ln/inbound_payment.rs
index cffb692d8..72f877978 100644
--- a/lightning/src/ln/inbound_payment.rs
+++ b/lightning/src/ln/inbound_payment.rs
@@ -129,9 +129,8 @@ fn min_final_cltv_expiry_delta_from_metadata(bytes: [u8; METADATA_LEN]) -> u16 {
 /// `ChannelManager` is required. Useful for generating invoices for [phantom node payments] without
 /// a `ChannelManager`.
 ///
-/// `keys` is generated by calling [`NodeSigner::get_inbound_payment_key`] and then
-/// calling [`ExpandedKey::new`] with its result. It is recommended to cache this value and not
-/// regenerate it for each new inbound payment.
+/// `keys` is generated by calling [`NodeSigner::get_inbound_payment_key`]. It is recommended to
+/// cache this value and not regenerate it for each new inbound payment.
 ///
 /// `current_time` is a Unix timestamp representing the current time.
 ///
```
